src/NServiceBus.Core/Pipeline/Incoming/LoadHandlersConnector.cs

```diff
@@ -51,12 +51,11 @@ public LoadHandlersConnector(MessageHandlerRegistry messageHandlerRegistry, ISyn
             }
         }
 
-        Task<CompletableSynchronizedStorageSession> AdaptOrOpenNewSynchronizedStorageSession(TransportTransaction transportTransaction, OutboxTransaction outboxTransaction, ContextBag contextBag)
+        async Task<CompletableSynchronizedStorageSession> AdaptOrOpenNewSynchronizedStorageSession(TransportTransaction transportTransaction, OutboxTransaction outboxTransaction, ContextBag contextBag)
         {
-            CompletableSynchronizedStorageSession session;
-            return adapter.TryAdapt(transportTransaction, out session) || adapter.TryAdapt(outboxTransaction, out session)
-                ? Task.FromResult(session)
-                : synchronizedStorage.OpenSession(contextBag);
+            return await adapter.TryAdapt(outboxTransaction, contextBag) 
```

Comment:
Not sure you need `await` here. If you do then `ConfigureAwait(false)`. But try returning directly